picasso/src/main/java/com/squareup/picasso/LruCache.java

```diff
@@ -19,159 +19,99 @@
 import android.graphics.Bitmap;
 import android.support.annotation.NonNull;
 
-import java.util.Iterator;
-import java.util.LinkedHashMap;
-import java.util.Map;
-
 import static com.squareup.picasso.Utils.KEY_SEPARATOR;
 
 /** A memory cache which uses a least-recently used eviction policy. */
-public class LruCache implements Cache {
-  final LinkedHashMap<String, BitmapAndSize> map;
-  private final int maxSize;
-
-  private int size;
-  private int putCount;
-  private int evictionCount;
-  private int hitCount;
-  private int missCount;
+public final class LruCache implements Cache {
+  final android.util.LruCache<String, LruCache.BitmapAndSize> cache;
 
   /** Create a cache using an appropriate portion of the available RAM as the maximum size. */
   public LruCache(@NonNull Context context) {
     this(Utils.calculateMemoryCacheSize(context));
   }
 
   /** Create a cache with a given maximum size in bytes. */
-  public LruCache(int maxSize) {
-    if (maxSize <= 0) {
-      throw new IllegalArgumentException("Max size must be positive.");
-    }
-    this.maxSize = maxSize;
-    this.map = new LinkedHashMap<>(0, 0.75f, true);
+  public LruCache(int maxByteCount) {
+    cache = new android.util.LruCache<String, LruCache.BitmapAndSize>(maxByteCount) {
+      @Override protected int sizeOf(String key, BitmapAndSize value) {
+        return (int) value.byteCount;
+      }
+    };
   }
 
   @Override public Bitmap get(@NonNull String key) {
-    if (key == null) {
-      throw new NullPointerException("key == null");
-    }
-
-    BitmapAndSize mapValue;
-    synchronized (this) {
-      mapValue = map.get(key);
-      if (mapValue != null) {
-        hitCount++;
-        return mapValue.bitmap;
-      }
-      missCount++;
-    }
-
-    return null;
+    BitmapAndSize bitmapAndSize = cache.get(key);
+    return bitmapAndSize != null ? bitmapAndSize.bitmap : null;
   }
 
   @Override public void set(@NonNull String key, @NonNull Bitmap bitmap) {
     if (key == null || bitmap == null) {
       throw new NullPointerException("key == null || bitmap == null");
     }
 
-    int addedSize = Utils.getBitmapBytes(bitmap);
-    if (addedSize > maxSize) {
-      return;
```

Comment:
this was a theoretical bug . . . replacing an existing key didn’t evict if the size was too big

Reply:
Are you sure? It was contributed because it was observed behavior.

Reply:
Oh sorry, I'm looking at the wrong method.

Reply:
Wait no I'm not... and you retained the behavior but added a comment?

Reply:
No, the “putting a giant image clears cache” is totally a real problem.

But there’s a theoretical problem with the fix. You have an image that’s beneath the limit in the cache. You try to update that with a higher-bytes copy of the same image. That doesn’t evict the lower-bytes predecessor, though it should.

Reply:
Ah, cool. 👍